// @shopify/shopify-app-remix/build/ts/server/unauthenticated/admin/unauthenticated.admin.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var data = {
    name: 'Unauthenticated admin',
    description: "Allows interacting with the Admin API when working outside of Shopify requests.\nThis enables apps to integrate with 3rd party services and perform background tasks.\n\n> Caution:\n> This function doesn't perform **any** validation and shouldn't rely on raw user input.\n\nWhen using this function, consider the following:\n\n#### Background tasks\n\nApps should ensure that the shop domain is authenticated when enqueueing jobs.\n\n#### 3rd party service requests\n\nApps must obtain the shop domain from the 3rd party service in a secure way.",
    category: 'Unauthenticated',
    type: 'object',
    isVisualComponent: false,
    definitions: [
        {
            title: 'unauthenticated.admin',
            description: 'Creates an unauthenticated Admin context.',
            type: 'GetUnauthenticatedAdminContext',
        },
    ],
    jsDocTypeExamples: ['UnauthenticatedAdminContext'],
    related: [
        {
            name: 'API context',
            subtitle: 'Interact with the Admin API.',
            url: '/docs/api/shopify-app-remix/apis/admin-api',
        },
    ],
};
exports.default = data;
