// @shopify/shopify-app-remix/build/ts/server/unauthenticated/storefront/unauthenticated.storefront.doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var data = {
    name: 'Unauthenticated storefront',
    description: "Allows interacting with the Storefront API when working outside of Shopify requests.\nThis enables apps to integrate with 3rd party services and perform background tasks.\n\n> Caution:\n> This function doesn't perform **any** validation and shouldn't rely on raw user input.\n\nWhen using this function, consider the following:\n\n#### Background tasks\n\nApps should ensure that the shop domain is authenticated when enqueueing jobs.\n\n#### 3rd party service requests\n\nApps must obtain the shop domain from the 3rd party service in a secure way.",
    category: 'Unauthenticated',
    type: 'object',
    isVisualComponent: false,
    definitions: [
        {
            title: 'unauthenticated.storefront',
            description: 'Creates an unauthenticated Storefront context.',
            type: 'GetUnauthenticatedStorefrontContext',
        },
    ],
    jsDocTypeExamples: ['UnauthenticatedStorefrontContext', 'StorefrontContext'],
    related: [
        {
            name: 'API context',
            subtitle: 'Interact with the Storefront API.',
            url: '/docs/api/shopify-app-remix/apis/storefront-api',
        },
    ],
};
exports.default = data;
