// @shopify/shopify-app-remix/build/ts/server/__test-helpers/setup-valid-session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setUpValidSession = void 0;
var tslib_1 = require("tslib");
var shopify_api_1 = require("@shopify/shopify-api");
var const_1 = require("./const");
function setUpValidSession(sessionStorage, sessionParams) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var overrides, id, session;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    overrides = {};
                    id = "offline_".concat(const_1.TEST_SHOP);
                    if (sessionParams === null || sessionParams === void 0 ? void 0 : sessionParams.isOnline) {
                        id = "".concat(const_1.TEST_SHOP, "_").concat(const_1.USER_ID);
                        // Expires one day from now
                        overrides.expires =
                            sessionParams.expires || new Date(Date.now() + 1000 * 3600 * 24);
                        overrides.onlineAccessInfo = {
                            associated_user_scope: 'testScope',
                            expires_in: 3600 * 24,
                            associated_user: {
                                id: const_1.USER_ID,
                                account_owner: true,
                                collaborator: true,
                                email: 'test@test.test',
                                email_verified: true,
                                first_name: 'Test',
                                last_name: 'User',
                                locale: 'en-US',
                            },
                        };
                    }
                    session = new shopify_api_1.Session(tslib_1.__assign({ id: id, shop: const_1.TEST_SHOP, isOnline: Boolean(sessionParams === null || sessionParams === void 0 ? void 0 : sessionParams.isOnline), state: 'test', accessToken: 'totally_real_token', scope: 'testScope' }, overrides));
                    return [4 /*yield*/, sessionStorage.storeSession(session)];
                case 1:
                    _a.sent();
                    return [2 /*return*/, session];
            }
        });
    });
}
exports.setUpValidSession = setUpValidSession;
