// @shopify/shopify-app-remix/build/cjs/server/types.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

// eslint-disable-next-line no-warning-comments
// TODO: Use this enum to replace the isCustomStoreApp config option in shopify-api-js
let AppDistribution = /*#__PURE__*/function (AppDistribution) {
  AppDistribution["AppStore"] = "app_store";
  AppDistribution["SingleMerchant"] = "single_merchant";
  AppDistribution["ShopifyAdmin"] = "shopify_admin";
  return AppDistribution;
}({});

// This can't be a type because it would reference itself
// eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style

let LoginErrorType = /*#__PURE__*/function (LoginErrorType) {
  LoginErrorType["MissingShop"] = "MISSING_SHOP";
  LoginErrorType["InvalidShop"] = "INVALID_SHOP";
  return LoginErrorType;
}({});

/**
 * An object your app can use to interact with Shopify.
 *
 * By default, the app's distribution is `AppStore`.
 */

exports.AppDistribution = AppDistribution;
exports.LoginErrorType = LoginErrorType;
